app/src/main/java/com/javahelp/frontend/gateway/RESTAPITokenGateway.java

```diff
@@ -0,0 +1,62 @@
+package com.javahelp.frontend.gateway;
+
+import com.javahelp.model.token.Token;
+
+import org.apache.hc.client5.http.async.methods.SimpleHttpRequest;
+import org.apache.hc.core5.concurrent.FutureCallback;
+
+import java.util.concurrent.Future;
+
+/**
+ * {@link RESTAPIGateway<T>} that makes use of {@link Token} authentication
+ *
+ * @param <T> type of response to expect
+ */
+public abstract class RESTAPITokenGateway<T> extends RESTAPIGateway<T> {
+
+    private static final String AUTH_TYPE = "JavaHelp";
+
+    IAuthInformationProvider provider;
+
+    /**
+     * Creates a new {@link RESTAPITokenGateway<T>}
+     *
+     * @param provider {@link IAuthInformationProvider} to use for authentication
+     */
+    public RESTAPITokenGateway(IAuthInformationProvider provider) {
+        this.provider = provider;
+    }
+
+    /**
+     * Makes a request using JavaHelp authorization.
+     * Overwrites existing Authorization headers
+     *
+     * @param request  {@link SimpleHttpRequest} to make
+     * @param callback {@link FutureCallback} to execute
+     * @return {@link Future} of {@link RESTAPIGatewayResponse<T>}
+     */
+    @Override
+    protected Future<RESTAPIGatewayResponse<T>> getResponse(SimpleHttpRequest request, FutureCallback<RESTAPIGatewayResponse<T>> callback) {
```

Comment:
What is Future?

Reply:
Its like a promise that the type parameter will exist in the future